lib/array.js

```diff
@@ -1,43 +1,43 @@
 'use strict';
 
 const common = require('@metarhia/common');
-
-// Asynchronous map (iterate parallel)
-//   items - <Array>, incoming
-//   fn - <Function>, to be executed for each value in the array
-//     current - <any>, current element being processed in the array
-//     callback - <Function>
-//       err - <Error> | <null>
-//       value - <any>
-//   done - <Function>, on done, optional
-//     err - <Error> | <null>
-//     result - <Array>
-const map = (items, fn, done) => {
-  done = done || common.emptyness;
-  const len = items.length;
-  if (!len) {
-    done(null, []);
+const { asyncIter } = require('./async-iterator.js');
+
+const map = (
+  // Asynchronous map (iterate parallel)
+  items, // array, incoming
+  fn, // function, (current, callback) => callback(err, value)
+  //   to be executed for each value in the array
+  //   current - current element being processed in the array
+  //   callback - function(err, value)
+  done // function (optional), on done callback function(err, result)
+) => {
+  if (!items[Symbol.iterator]) {
+    done(null, null);
     return;
   }
-  let errored = false;
-  let count = 0;
-  const result = new Array(len);
-
-  const next = (index, err, value) => {
-    if (errored) return;
-    if (err) {
-      errored = true;
-      done(err);
-      return;
-    }
-    result[index] = value;
-    count++;
-    if (count === len) done(null, result);
-  };
-
-  for (let i = 0; i < len; i++) {
-    fn(items[i], next.bind(null, i));
-  }
+  done = done || common.emptyness;
+  const isArray = Array.isArray(items);
+  const data = asyncIter(items).map(
+    item =>
+      new Promise((resolve, reject) => {
+        fn(item, (err, res) => {
+          if (err) reject(err);
+          else resolve(res);
+        });
+      })
+  );
+  const promise =
+    isArray || items.constructor.name === 'String'
+      ? data.toArray()
+      : data.collectTo(items.constructor);
```

Comment:
After consulting with @belochub, we agreed that we should return an `Iterator` object if `items` is not an `Array`.